ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py

```diff
@@ -93,3 +96,151 @@ def from_shapefile_to_poly(shapefile: Path, out_path: Path, outname: str = ""):
+    """
+        Based on osmnx.
+    """
+    node_coor = (node[1]['x'], node[1]['y'])
+    # get u, v, key, geom from all the graph edges
+    gdf_edges = graph_to_gdf(graph, nodes=False, fill_edge_geometry=True)
+    edges = gdf_edges[["u", "v", "data", "key", "geometry"]].values
+
+    # convert lat,lng (y,x) node to x,y for shapely distance operation
+    xy_point = Point(node_coor)
+
+    # calculate Euclidean distance from each edge's geometry to this node
+    edge_distances = [(edge, xy_point.distance(edge[4])) for edge in edges]
+
+    # the nearest edge minimizes the non-zero distance to the node
+    (u, v, data, key, geom), dist = min(edge_distances, key=lambda x: x[1] if x[1] > 0 else float('inf'))
+    utils.log(f"Found nearest edge ({u, v, data, key}) to node {node}")
+
+    # return results requested by caller
+    if data == 0 and key == 0:
+        if return_dist and return_geom:
+            return {
+                "node": node,
+                "nearest_edge": (u, v, geom, dist)}
+        elif return_dist:
+            return {
+                "node": node,
+                "nearest_edge": (u, v, dist)}
+        elif return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, geom)}
+        else:
+            return {"node": node,
+                    "nearest_edge": (u, v)}
+    elif data == 0:
+        if return_dist and return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, key, geom, dist)}
+        elif return_dist:
+            return {"node": node,
+                    "nearest_edge": (u, v, key, dist)}
+        elif return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, key, geom)}
+        else:
+            return {"node": node,
+                    "nearest_edge": (u, v, key)}
+    elif key == 0:
+        if return_dist and return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, geom, dist)}
+        elif return_dist:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, dist)}
+        elif return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, geom)}
+        else:
+            return {"node": node,
+                    "nearest_edge": (u, v, data)}
+    else:
+        if return_dist and return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key, geom, dist)}
+        elif return_dist:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key, dist)}
+        elif return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key, geom)}
+        else:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key)}
+
+
+def _is_endnode_simplified(graph: MultiDiGraph, node: int) -> bool:
```

Comment:
This method should not be with  underscore `_`. This symbol represents that the method is private and should only be accessed by methods within the same context, however you are importing it from another module. 

Please do not forget to describe the docstrings.